### ISSUE TITLE
torchvision.io.read_image : [Errno 42] Illegal byte sequence

### ISSUE THREAD

**Issue by m-k-moriya** [pytorch/vision#2931]
The image cannot be loaded if the path contains multibyte strings.

## To Reproduce
``` python
import torchvision

file_path = "D:\\日本語(Japanese)\\Sample.png"

# It's work.
f = open(file_path, 'rb')
f.close()

# It's not work.
tensor_image = torchvision.io.read_image(file_path)

--------------------------------------------------------
data = torch.ops.image.read_file(path)
RuntimeError: [Errno 42] Illegal byte sequence: 'D:\日本語(Japanese)\Sample.png'
```

## Environment
PyTorch / torchvision Version (e.g., 1.0 / 0.4.0): 1.7.0/0.8.1
OS (e.g., Linux): Windows 10
How you installed PyTorch / torchvision (conda, pip, source): pip
Build command you used (if compiling from source):
Python version: 3.7.9
CUDA/cuDNN version: 10.1
GPU models and configuration: Geforce GTX 1060
Any other relevant information: None

cc @peterjc123 @nbcsm @guyang3532 @maxluk @gunandrose4u @smartcat2010 @mszhanyi

**Comment by andfoy** [comment]
Hi @m-k-moriya, thanks for opening this issue, what happens if the image only has ascii characters?

**Comment by fmassa** [comment]
To complement @andfoy message, could you also try doing
```python
file_path = "D:\\日本語(Japanese)\\Sample.png"

with open(file_path, 'rb') as f
    raw_data = f.read()
    raw_data_tensor = torch.as_tensor(list(raw_data), dtype=torch.uint8)

tensor_image = torchvision.io.decode_image(raw_data_tensor)
```
so that we can try to identify more easily if the issue is with `read_file` or with `decode_image`?

**Comment by m-k-moriya** [comment]
@andfoy , @fmassa, thank you for reply.

The code below works fine.

``` python
import torch
import torchvision

file_path = "D:\\日本語(Japanese)\\Sample.png"

with open(file_path, 'rb') as f:
    raw_data = f.read()
    raw_data_tensor = torch.as_tensor(list(raw_data), dtype=torch.uint8)

tensor_image = torchvision.io.decode_image(raw_data_tensor)
print(tensor_image)

---------
tensor([[[ 94,  95,  88,  ..., 192, 195, 196],
         [ 97,  93,  88,  ..., 195, 192, 195],
         [ 94,  95,  88,  ..., 196, 194, 194],
         ...,
```

**Comment by m-k-moriya** [comment]
@andfoy , The code below that only has ascii characters works fine.

``` python
import torchvision

file_path = "D:\\Japanese\\Sample.png"

tensor_image = torchvision.io.read_image(file_path)
print(tensor_image)
--------
tensor([[[ 94,  95,  88,  ..., 192, 195, 196],
         [ 97,  93,  88,  ..., 195, 192, 195],
         [ 94,  95,  88,  ..., 196, 194, 194],
```

**Comment by fmassa** [comment]
Thanks a lot for confirming @m-k-moriya !

This means there is an issue in `read_file` (at least on Windows) for non-ascii characters. We will be investigating it.

**Comment by m-k-moriya** [comment]
@fmassa , Thank you for your investigation.
I would appreciate your continuous support.

**Comment by fmassa** [comment]
I've looked a bit into it, and this looks like an issue that only happens on Windows, see https://github.com/pytorch/vision/pull/2946

@peterjc123 do you have an idea of what we might need to change in `read_file` and `write_file`?

**Comment by peterjc123** [comment]
@fmassa Yeah, but the problem is not so simple. Since the implemention of `read_file` relies on `THMapAllocator`, we'll need to replace all the ascii variants of Windows API with the unicode ones in all the related codepaths. And that's a lot work to do.

**Comment by fmassa** [comment]
@peterjc123 so you are saying that the problem is within `torch.from_file`, and not the file size reading?
If that's the case then I'd say it's a pretty serious bug in `torch.from_file` that should be fixed in PyTorch core.

**Comment by peterjc123** [comment]
@fmassa I mean both. File size reading is easy to fix comparing with the one related to PyTorch.

> If that's the case then I'd say it's a pretty serious bug in `torch.from_file` that should be fixed in PyTorch core.

Yeah, maybe you can create a new issue in PyTorch.

**Comment by fmassa** [comment]
@peterjc123 you are right, the second commit in #2946 adds a new test to see if `torch.from_file` also suffers from this issue, and that is indeed the case.
So the solution is more complicated than I originally thought if we want to use the `from_file` from PyTorch.

I think a workaround would be to dispatch to a custom reading implementation on Windows, so that we don't do through `torch.from_file`, and create an issue in PyTorch to track this down so that in the future this can be fixed in PyTorch itsefl.

Thoughts?

**Comment by peterjc123** [comment]
@fmassa That sounds good to me. BTW, do you have a reference implementation for it?

**Comment by andfoy** [comment]
I think we could use ctypes and calls to ReadFileW

![imagen](https://user-images.githubusercontent.com/1878982/98016333-d55adb00-1dcb-11eb-9ba6-091495dd3642.png)

**Comment by fmassa** [comment]
@andfoy this implementation wouldn't allow for torchscript support, so we need to keep it entirely in C++ for now.

@peterjc123 not right now, I was thinking about doing something very basic that involves reading into a vector via `ifstream` and then converting to a Tensor via `from_blob`, which is btw very similar to what you did in your PR

**Comment by fmassa** [comment]
@m-k-moriya This issue has been fixed by @peterjc123 in #2949, and will be present in the next torchvision nightly, as well as in the next minor release which will be out in December.

Thanks for flagging this bug!

**Comment by m-k-moriya** [comment]
@fmassa , @peterjc123, @andfoy, Thank you for your prompt response!
I'm looking forward to the next release.
